[AArch64][SME] Support split ZPR and PPR area allocation

MacDue opened the pull request:
For a while we have supported the `-aarch64-stack-hazard-size=<size>` option, which adds "hazard padding" between GPRs and FPR/ZPRs. However, there is currently a hole in this mitigation as PPR and FPR/ZPR accesses to the same area also cause streaming memory hazards (this is noted by `-pass-remarks-analysis=sme -aarch64-stack-hazard-remark-size=<val>`), and the current stack layout places PPRs and ZPRs within the same area.

Which looks like:

```
------------------------------------  Higher address
| callee-saved gpr registers        |
|---------------------------------- |
| lr,fp  (a.k.a. "frame record")    |
|-----------------------------------| <- fp(=x29)
|   <hazard padding>                |
|-----------------------------------|
| callee-saved fp/simd/SVE regs     |
|-----------------------------------|
|        SVE stack objects          |
|-----------------------------------|
| local variables of fixed size     |
|   <FPR>                           |
|   <hazard padding>                |
|   <GPR>                           |
------------------------------------| <- sp
                                    | Lower address
```

With this patch the stack (and hazard padding) is rearranged so that hazard padding is placed between the PPRs and ZPRs rather than within the (fixed size) callee-save region. Which looks something like this:

```
------------------------------------  Higher address
| callee-saved gpr registers        |
|---------------------------------- |
| lr,fp  (a.k.a. "frame record")    |
|-----------------------------------| <- fp(=x29)
|        callee-saved PPRs          |
|        PPR stack objects          | (These are SVE predicates)
|-----------------------------------|
|   <hazard padding>                |
|-----------------------------------|
|       callee-saved ZPR regs       | (These are SVE vectors)
|        ZPR stack objects          | Note: FPRs are promoted to ZPRs
|-----------------------------------|
| local variables of fixed size     |
|   <FPR>                           |
|   <hazard padding>                |
|   <GPR>                           |
------------------------------------| <- sp
                                    | Lower address
```

This layout is only enabled if:

 * SplitSVEObjects are enabled (`-aarch64-split-sve-objects`)
   - (This may be enabled by default in a later patch)
 * Streaming memory hazards are present
   - (`-aarch64-stack-hazard-size=<val>` != 0)
 * PPRs and FPRs/ZPRs are on the stack
 * There's no stack realignment or variable-sized objects
   - This is left as a TODO for now

Additionally, any FPR callee-saves that are present will be promoted to ZPRs. This is to prevent stack hazards between FPRs and GRPs in the fixed size callee-save area (which would otherwise require more hazard padding, or moving the FPR callee-saves).

This layout should resolve the hole in the hazard padding mitigation, and is not intended change codegen for non-SME code.

efriedma-quic commented:
In the implementation you're interested in, is there not a hazard between PPRs and GPRs?

What's the interaction between this and aarch64-enable-zpr-predicate-spills?

MacDue commented:
> In the implementation you're interested in, is there not a hazard between PPRs and GPRs?

There are no hazards between PPRs and GPRs (those types of memory accesses can both be considered as occurring on the CPU). 

> What's the interaction between this and aarch64-enable-zpr-predicate-spills?

It cannot be enabled at the same time. We intend to remove `aarch64-enable-zpr-predicate-spills` after this lands (as it was a stopgap solution).

efriedma-quic reviewed (commented):
If you express the size of the hazard padding between the PPRs and ZPRs as a scalable size, that might simplify some of the logic?  You wouldn't need to represent the two areas as separate stacks, at least.

Maybe it's also worth considering "preallocating" some stack slots, along the lines of LocalStackSlotAllocation; anything we can allocate in an earlier pass doesn't have to be part of the core stack layout, and we probably get better code if we have a "base" for ZPR variables.

Do we care at all how stack protectors fit into all of this?  I suspect the stack protector ends up in an unfortunate position by default.

efriedma-quic commented on llvm/lib/Target/AArch64/AArch64FrameLowering.h:
```diff
@@ -57,7 +57,8 @@ class AArch64FrameLowering : public TargetFrameLowering {
                                           bool isSVE, Register &FrameReg,
-                                          bool PreferFP, bool ForSimm) const;
+                                          bool PreferFP, bool ForSimm,
+                                          int64_t FI = -1) const;
```
resolveFrameOffsetReference only has three callers; I'm not sure you're getting much by using a default argument here.

MacDue commented:
> If you express the size of the hazard padding between the PPRs and ZPRs as a scalable size, that might simplify some of the logic? You wouldn't need to represent the two areas as separate stacks, at least.

It would, but for the sizes of hazard padding and vscale we're interested in, it would result in a much larger allocation than necessary and likely complicate addressing predicates and vectors more so (due to limited ranges for scalable offsets).

MacDue commented on llvm/lib/Target/AArch64/AArch64FrameLowering.cpp:
```diff
@@ -3780,25 +3938,49 @@ void AArch64FrameLowering::determineStackHazardSlot(
+      HasFPRStackObjects |= SlotTypes[FI] == SlotType::ZPRorFPR;
+      // For SplitSVEObjects remember that this stack slot is a predicate, this
+      // will be needed later when determining the frame layout.
+      if (SlotTypes[FI] == SlotType::PPR) {
```
Which bit does not look right? `SlotType::PPR`  (not to be confused with `SlotType::GPR`) is only set if the original stack ID was scalable and all accesses to that slot used predicate load/stores. The original stack ID could be `ScalableVector`, as the earlier selection of the stack ID is only based on the type size. Therefore, we change it to `ScalablePredVector` here so that it can be sorted into the correct region.

efriedma-quic replied:
Oh, I got confused by this bit:

```
        if (IsScalable || AArch64InstrInfo::isFpOrNEON(MI)) {
          SlotTypes[*FI] |= IsPPR ? SlotType::PPR : SlotType::ZPRorFPR;
        } else {
          SlotTypes[*FI] |= SlotType::GPR;
        }
```

Maybe rewrite it as the following, to split the cases a bit more cleanly?

```
if (IsScalable) {
  SlotTypes[*FI] |= isPPRAccess(MI) ? SlotType::PPR : SlotType::ZPRorFPR;
} else {
  SlotTypes[*FI] |= AArch64InstrInfo::isFpOrNEON(MI) ? SlotType::ZPRorFPR : SlotType::GPR;
}
```

Also, I'm not sure isFpOrNEON is quite sufficient here; we probably also want to handle SVE accesses to fixed-width slots.

MacDue replied:
> Maybe rewrite it as the following, to split the cases a bit more cleanly?

:+1: I've tidied up this case.

> Also, I'm not sure isFpOrNEON is quite sufficient here; we probably also want to handle SVE accesses to fixed-width slots.

Hmmm, I'll look into this, these checks are mostly the same as the for the current hazard padding option.

MacDue committed:
Small fixups

Change-Id: I64f503597717441461ede15dc50709f9af498288

MacDue commented on llvm/lib/Target/AArch64/AArch64PrologueEpilogue.cpp:
```diff
@@ -1405,111 +1432,186 @@ void AArch64EpilogueEmitter::emitEpilogue() {
+          // to compute the base address by subtracting the offest in a
+          // temporary register first (to avoid briefly deallocating the SVE
+          // CS).
+          CalleeSaveBase = MBB.getParent()->getRegInfo().createVirtualRegister(
+              &AArch64::GPR64RegClass);
```
This code is just moved (and not reused for the split-SVE implementation), so I don't think refactoring it would be relevant for this PR.

MacDue committed:
Add stack layout comment

Change-Id: If04156b0ef35e89a517d8bc06f121d2555ee9379

MacDue committed:
[AArch64][SME] Reshuffle emit[prologue|epilogue]() for splitSVEObjects (NFCI)

Requested in https://github.com/llvm/llvm-project/pull/142392#discussion_r2207880079

Change-Id: I842faddea1bd54c5e30a9985782baf5dce37e5bb

MacDue committed:
[AArch64][SME] Support split ZPR and PPR area allocation

For a while we have supported the `-aarch64-stack-hazard-size=<size>`
option, which adds "hazard padding" between GPRs and FPR/ZPRs. However,
there is currently a hole in this mitigation as PPR and FPR/ZPR accesses
to the same area also cause streaming memory hazards (this is noted by
`-pass-remarks-analysis=sme -aarch64-stack-hazard-remark-size=<val>`),
and the current stack layout places PPRs and ZPRs within the same area.

Which looks like:

------------------------------------  Higher address
| callee-saved gpr registers        |
|---------------------------------- |
| lr,fp  (a.k.a. "frame record")    |
|-----------------------------------| <- fp(=x29)
|   <hazard padding>                |
|-----------------------------------|
| callee-saved fp/simd/SVE regs     |
|-----------------------------------|
|        SVE stack objects          |
|-----------------------------------|
| local variables of fixed size     |
|   <FPR>                           |
|   <hazard padding>                |
|   <GPR>                           |
------------------------------------| <- sp
                                    | Lower address

With this patch the stack (and hazard padding) is rearranged so that
hazard padding is placed between the PPRs and ZPRs rather than within
the (fixed size) callee-save region. Which looks something like this:

------------------------------------  Higher address
| callee-saved gpr registers        |
|---------------------------------- |
| lr,fp  (a.k.a. "frame record")    |
|-----------------------------------| <- fp(=x29)
|        callee-saved PPRs          |
|        PPR stack objects          | (These are SVE predicates)
|-----------------------------------|
|   <hazard padding>                |
|-----------------------------------|
|       callee-saved ZPR regs       | (These are SVE vectors)
|        ZPR stack objects          | Note: FPRs are promoted to ZPRs
|-----------------------------------|
| local variables of fixed size     |
|   <FPR>                           |
|   <hazard padding>                |
|   <GPR>                           |
------------------------------------| <- sp
                                    | Lower address

This layout is only enabled if:

 * SplitSVEObjects are enabled (`-aarch64-split-sve-objects`)
   - (This may be enabled by default in a later patch)
 * Streaming memory hazards are present
   - (`-aarch64-stack-hazard-size=<val>` != 0)
 * PPRs and FPRs/ZPRs are on the stack
 * There's no stack realignment or variable-sized objects
   - This is left as a TODO for now

Additionally, any FPR callee-saves that are present will be promoted to
ZPRs. This is to prevent stack hazards between FPRs and GRPs in the
fixed size callee-save area (which would otherwise require more hazard
padding, or moving the FPR callee-saves).

This layout should resolve the hole in the hazard padding mitigation,
and is not intended change codegen for non-SME code.

Change-Id: I2e1906577c2ac79c40bc69e7c15e3ef09857445f

MacDue commented on llvm/lib/Target/AArch64/AArch64FrameLowering.cpp:
```diff
@@ -2324,73 +2340,96 @@ void AArch64FrameLowering::emitPrologue(MachineFunction &MF,
-                       NeedsWinCFI, &HasWinCFI, EmitAsyncCFI && !HasFP,
-                       CFAOffset, MFI.hasVarSizedObjects());
+  if (AFI->hasSplitSVEObjects()) {
+    assert(!FPAfterSVECalleeSaves &&
```
Done https://github.com/llvm/llvm-project/pull/161217 (and adjusted the PR stack).

MacDue committed:
Don't use setIsSVECC()

Change-Id: I58542d83ae47b292ba3fa15f47ca0b9cb5a16141

MacDue commented on llvm/lib/Target/AArch64/AArch64FrameLowering.cpp:
```diff
@@ -2277,6 +2351,70 @@ void AArch64FrameLowering::determineStackHazardSlot(
+
+    // FIXME: Avoid setting setting the CC. Since we've replaced FPRs with ZPRs
+    // we need to set this or later in PEI the ZPR CS will be masked out.
+    AFI->setIsSVECC(true);
```
Done, I've also added a check that splitSVEObjects is only applied to functions with `C`, `Fast`, or `SVE` calling conventions, as otherwise promoting FPRs to ZPRs won't work (as the `SaveList` will be overridden by some other calling convention).

MacDue committed:
Remove unused iterators

Change-Id: I2c0e8290e4c224045aa12031cd966bc66bd84ac4

MacDue committed:
Add iterator changes from #161217

Change-Id: I2703f109fda15a437bd918fbd94a97566f26e161

sdesmalen-arm reviewed (commented):
Mostly some minor nits and request for some more tests, but otherwise it looks good to me.

MacDue committed:
Fixups

Change-Id: I44aaa861c82fa44fe4c762366572a7367e8bf5c0

MacDue commented on llvm/test/CodeGen/AArch64/split-sve-stack-frame-layout.ll:
```diff
@@ -0,0 +1,750 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc < %s -mtriple=aarch64 -mattr=+sve2 -aarch64-stack-hazard-in-non-streaming -aarch64-split-sve-objects -aarch64-streaming-hazard-size=1024 | FileCheck %s
+; RUN: llc < %s -mtriple=aarch64 -mattr=+sve2 -aarch64-stack-hazard-in-non-streaming -aarch64-split-sve-objects -aarch64-streaming-hazard-size=1024 -pass-remarks-analysis=stack-frame-layout 2>&1 >/dev/null | FileCheck %s --check-prefixes=CHECK-FRAMELAYOUT
+
```
Added tests :+1:

MacDue committed:
Tweak comment layout

Change-Id: I048bf702e3a507275bb503217e7b91b2d0d67498

MacDue committed:
Tweak comment layout x2

Change-Id: I0479ebfbfd33a96bb7b50ca6c7b691f594383053

sdesmalen-arm reviewed (approved):
Thanks @MacDue, LGTM!
